### PR TITLE
feat: add roles to control user ui route

### PR DESCRIPTION
**Pull request by tyyzqmf**
----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

## Summary

1. **backend:** Add apis for user management
2. **frontend:** Add `UserContext` and `RoleRoute` to control page route auth

### Roles:
- **Admin**: Access all pages
- **Operator**: Only access pipeline pages
- **Analyst**: Only access explore pages

### Navigation:
- **Admin**: All items (include new item `System/User`)
- **Operator**: exclude new item `System/User` and `Explore`
- **Analyst**: All items of explore

截屏2023-08-31 22 03 46

### Login Callback:
- **Admin/Operator**: redirect to `'/'`
- **Analyst**: redirect to `'/analytics'`

### 401 page
- If Developer access `'/analytics/*'` pages show 401
- If Analyst access **NOT** `'/analytics/*'` pages show 401

截屏2023-08-31 22 04 10

## Test checklist

- [x] add new test cases
- [ ] all code changes are covered by unit tests
- [ ] end-to-end tests
  - [ ] deploy web console with CloudFront + S3 + API gateway
  - [ ] deploy web console within VPC
  - [ ] deploy ingestion server
    - [ ] with MSK sink
    - [ ] with KDS sink
    - [ ] with S3 sink
  - [ ] deploy data processing
  - [ ] deploy data modeling
    - [ ] new Redshift Serverless
    - [ ] provisioned Redshift
    - [ ] Athena
  - [ ] deploy with reporting

## Is it a breaking change

- [ ] add parameters without default value in stack
- [ ] introduce new service permission in stack
- [ ] introduce new top level stack module

## Miscellaneous

- [ ] introduce new symbol link source file(s) to be shared among infra code, web console frontend, and web console backend

### CODE REVIEW

File: frontend/src/components/common/RoleRoute.tsx [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+import { UserContext } from 'context/UserContext';
+import AccessDenied from 'pages/error-page/AccessDenied';
+import React, { useContext } from 'react';
+import { IUserRole } from 'ts/const';
+
+const RoleRoute = ({
+  children,
+  roles,
+}: {
+  children: React.JSX.Element;
```
Comment (chenhaiyun) [inline frontend/src/components/common/RoleRoute.tsx:23]: Change the children type to `ReactElement`

File: frontend/src/App.tsx [hunk 2]
```diff
@@ -37,17 +40,24 @@ import PluginList from 'pages/plugins/PluginList';
...
+import { CONFIG_URL, IUserRole, PROJECT_CONFIG_JSON } from 'ts/const';
 import Home from './pages/home/Home';
 
 const LoginCallback: React.FC = () => {
+  const currentUser = useContext(UserContext);
+
   useEffect(() => {
     const baseUrl = '/';
-    window.location.href = baseUrl;
+    if (currentUser?.role === IUserRole.ANALYST) {
```
Comment (chenhaiyun) [inline frontend/src/App.tsx:56]: If user refresh the page (F5), the page will always redirect `/` or `/analytics`. 
For example: When user in pipeline detail page. they want to refresh the data and refresh...

File: frontend/src/components/layouts/Navigation.tsx [hunk 3]
```diff
@@ -44,22 +51,34 @@ const Navigation: React.FC<INavigationProps> = (props: INavigationProps) => {
...
       type: 'link',
       text: t('nav.doc'),
-      href: getDoucmentList(i18n.language),
+      href: getDocumentList(i18n.language),
       external: true,
     },
   ];
-  const navAnalyticsItems: SideNavigationProps.Item[] = [
+  const analyticsNavItems: SideNavigationProps.Item[] = [
     // {
```
Comment (chenhaiyun) [inline frontend/src/components/layouts/Navigation.tsx:82]: Remove the comments if not use anymore..

File: frontend/src/components/layouts/Header.tsx [hunk 4]
```diff
@@ -138,10 +141,12 @@ const Header: React.FC<IHeaderProps> = (props: IHeaderProps) => {
                   onItemClick: (item) => {
                     if (item.detail.id === 'signout') {
                       if (fullLogoutUrl) {
-                        signOut && signOut();
+                        if (signOut) {
```
Comment (chenhaiyun) [inline frontend/src/components/layouts/Header.tsx:144]: May be we can change to `signOut?.()`

File: frontend/src/pages/error-page/AccessDenied.tsx [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+import { Alert, Container, ContentLayout } from '@cloudscape-design/components';
+import React from 'react';
+import { useTranslation } from 'react-i18next';
+
+const AccessDenied: React.FC = () => {
+  const { t } = useTranslation();
+  return (
+    <ContentLayout>
+      <Container>
+        <div className="">
```
Comment (chenhaiyun) [inline frontend/src/pages/error-page/AccessDenied.tsx:23]: Add a class `mt-10` to make some space

File: frontend/src/pages/user/UserList.tsx [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+    },
+    {
+      id: 'name',
+      header: t('user:labels.tableColumnName'),
+      cell: (e: { name: string }) => {
+        return e.name;
+      },
+      minWidth: 180,
+      editConfig: {
+        ariaLabel: 'Edit display name',
```
Comment (chenhaiyun) [inline frontend/src/pages/user/UserList.tsx:49]: Change these to i18n or remove this?

File: frontend/src/pages/user/UserList.tsx [hunk 7]
```diff
@@ -0,0 +1,202 @@
...
+        ) => {
+          return (
+            <Input
+              autoFocus={true}
+              ariaLabel="Edit user name"
+              value={currentValue ?? item.name}
+              onChange={(item) => {
+                setValue(item.detail.value);
+              }}
+              placeholder="Enter user name"
```
Comment (chenhaiyun) [inline frontend/src/pages/user/UserList.tsx:69]: i18n

File: frontend/src/pages/user/UserList.tsx [hunk 8]
```diff
@@ -0,0 +1,202 @@
...
+      },
+      minWidth: 180,
+      editConfig: {
+        ariaLabel: 'Edit display name',
+        errorIconAriaLabel: 'Display Name Validation Error',
+        editIconAriaLabel: 'editable',
+        validation(item: any, value: any) {
+          return !new RegExp(XSS_PATTERN).test(value)
+            ? undefined
+            : 'Invalid input';
```
Comment (chenhaiyun) [inline frontend/src/pages/user/UserList.tsx:55]: i18n

File: frontend/src/pages/user/UserList.tsx [hunk 9]
```diff
@@ -0,0 +1,202 @@
...
+          );
+        },
+      },
+    },
+    {
+      id: 'role',
+      header: t('user:labels.tableColumnRole'),
+      minWidth: 200,
+      editConfig: {
+        ariaLabel: 'Edit state',
```
Comment (chenhaiyun) [inline frontend/src/pages/user/UserList.tsx:80]: Change these to i18n or remove this?

Comment (chenhaiyun) [reply frontend/src/pages/user/UserList.tsx:80]: I think this can be as a common config?

File: frontend/src/pages/user/UserTable.tsx [hunk 10]
```diff
@@ -0,0 +1,237 @@
...
+    onSortingChange: (event: any) => {
+      if (collectionProps.onSortingChange) {
+        collectionProps.onSortingChange(event);
+      }
+      persistChanges();
+    },
+  };
+
+  const handleSubmit = async (currentItem: IUser, column: any, value: any) => {
+    // if (currentItem.operator === 'Clickstream') {
```
Comment (chenhaiyun) [inline frontend/src/pages/user/UserTable.tsx:143]: Remove the comments if not use.

File: frontend/src/pages/user/UserTableHeader.tsx [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+  onInfoLinkClick,
+  ...props
+}: UserTableHeaderProps) {
+  return (
+    <Header
+      variant="awsui-h1-sticky"
+      info={
+        onInfoLinkClick && (
+          <Link variant="info" onFollow={onInfoLinkClick}>
+            Info
```
Comment (chenhaiyun) [inline frontend/src/pages/user/UserTableHeader.tsx:43]: i18n

File: frontend/src/components/common/RoleRoute.tsx (hunk 1 above)
Comment (tyyzqmf) [reply frontend/src/components/common/RoleRoute.tsx:23]: done

File: frontend/src/App.tsx (hunk 2 above)
Comment (tyyzqmf) [reply frontend/src/App.tsx:56]: The `useEffect` in **LoginCallback**, so it has no impact on refresh.

File: src/control-plane/backend/click-stream-api.ts [hunk 12]
```diff
@@ -390,12 +445,6 @@ export class ClickStreamApiConstruct extends Construct {
       ...apiFunctionProps,
     });
 
-    dictionaryTable.grantReadWriteData(this.clickStreamApiFunction);
```
Comment (zxkane) [inline src/control-plane/backend/click-stream-api.ts:393]: What's the benefit to replace those calls with explitly creating role?

File: src/control-plane/backend/click-stream-api.ts [hunk 13]
```diff
@@ -170,13 +173,30 @@ export class ClickStreamApiConstruct extends Construct {
       },
     });
 
+    const userTable = new Table(this, 'ClickstreamUser', {
+      partitionKey: {
+        name: 'email',
```
Comment (zxkane) [inline src/control-plane/backend/click-stream-api.ts:178]: Might `uid` be better? We might can not get email from some OIDC providers.

File: src/control-plane/backend/insert-admin-user.ts [hunk 14]
```diff
@@ -0,0 +1,121 @@
...
+
+  readonly crForAddAdminUser: CustomResource;
+
+  constructor(scope: Construct, id: string, props: AddAdminUserProps) {
+    super(scope, id);
+
+    this.crForAddAdminUser = this.createAddAdminUserCustomResource(props);
+  }
+
+  private createAddAdminUserCustomResource(props: AddAdminUserProps): CustomResource {
```
Comment (zxkane) [inline src/control-plane/backend/insert-admin-user.ts:43]: Using too much sketching code for making a simple API call, try to use CDK's `AwsCustomResource` to simplify it

https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.custom_resources-readme.html#custom-resources-for-aws-apis

**Review by zxkane — CHANGES_REQUESTED** [review]

File: src/control-plane/backend/insert-admin-user.ts [hunk 15]
```diff
@@ -0,0 +1,66 @@
...
+
+export class AddAdminUser extends Construct {
+
+  readonly crForAddAdminUser: AwsCustomResource;
+
+  constructor(scope: Construct, id: string, props: AddAdminUserProps) {
+    super(scope, id);
+    const awsSdkCall: AwsSdkCall = {
+      service: 'DynamoDB',
+      action: 'putItem',
```
Comment (zxkane) [inline src/control-plane/backend/insert-admin-user.ts:34]: We might use `updateItem` to upsert the admin user.

File: src/control-plane/backend/insert-admin-user.ts [hunk 16]
```diff
@@ -0,0 +1,66 @@
...
+          createAt: { N: Date.now().toString() },
+          updateAt: { N: Date.now().toString() },
+          operator: { S: 'Clickstream' },
+          deleted: { BOOL: false },
+        },
+        ConditionExpression: 'attribute_not_exists(uid)',
+      },
+    };
+
+    this.crForAddAdminUser = new AwsCustomResource(this, 'AddAdminUserAwsCustomResource', {
```
Comment (zxkane) [inline src/control-plane/backend/insert-admin-user.ts:50]: Also, add `onDelete` for the case updating admin user.

**Review by zxkane — CHANGES_REQUESTED** [review]

File: src/control-plane/backend/insert-admin-user.ts (hunk 15 above)
Comment (tyyzqmf) [reply src/control-plane/backend/insert-admin-user.ts:34]: uid is PK, so can not update. There will append new admin user to DDB if stack update email.

File: src/control-plane/backend/insert-admin-user.ts (hunk 16 above)
Comment (tyyzqmf) [reply src/control-plane/backend/insert-admin-user.ts:50]: added.

File: src/control-plane/backend/insert-admin-user.ts (hunk 15 above)
Comment (tyyzqmf) [reply src/control-plane/backend/insert-admin-user.ts:34]: done.

**Review by zxkane — APPROVED** [review]

**Pull request merged**